packages/gatsby/src/utils/api-runner-node.js

```diff
@@ -179,151 +179,63 @@ const runAPI = (plugin, api, args) => {
+  // "FAKE_API_CALL" is used when code needs to trigger something to happen
+  // once the the API queue is empty. Ideally of course we'd have an API
+  // (returning a promise) for that. But this works nicely in the meantime.
+  if (!apiList[api] && api !== `FAKE_API_CALL`) {
+    reporter.panic(`api: "${api}" is not a valid Gatsby api`)
+  }
 
-    apisRunningById.set(apiRunInstance.id, apiRunInstance)
-    if (apisRunningByTraceId.has(apiRunInstance.traceId)) {
-      const currentCount = apisRunningByTraceId.get(apiRunInstance.traceId)
-      apisRunningByTraceId.set(apiRunInstance.traceId, currentCount + 1)
-    } else {
-      apisRunningByTraceId.set(apiRunInstance.traceId, 1)
-    }
+  const { store } = require(`../redux`)
+  const plugins = store.getState().flattenedPlugins
+  // Get the list of plugins that implement gatsby-node.
+  if (!filteredPlugins) {
+    filteredPlugins = plugins.filter(plugin => hasAPIFile(plugin))
+  }
 
-    Promise.mapSeries(noSourcePluginPlugins, plugin => {
-      let pluginName =
+  // Break infinite loops.
+  // Sometimes a plugin will implement an API and call an action which will
+  // trigger the same API being called. `onCreatePage` is the only example
+  // right now. In these cases, we should avoid calling the originating plugin
+  // again.
+  const noSourcePluginPlugins = pluginSource
+    ? filteredPlugins.filter(p => p.name !== pluginSource)
+    : filteredPlugins
+
+  const results = []
+  for (const plugin of noSourcePluginPlugins) {
+    try {
+      const result = await runAPI(plugin, api, { ...args, parentSpan: apiSpan })
+      results.push(result)
+    } catch (err) {
+      const pluginName =
         plugin.name === `default-site-plugin`
           ? `gatsby-node.js`
           : `Plugin ${plugin.name}`
+      reporter.panicOnBuild(`${pluginName} returned an error`, err)
+    }
+  }
 
-      return new Promise(resolve => {
-        resolve(runAPI(plugin, api, { ...args, parentSpan: apiSpan }))
-      }).catch(err => {
-        reporter.panicOnBuild(`${pluginName} returned an error`, err)
-        return null
-      })
-    }).then(results => {
-      // Remove runner instance
-      apisRunningById.delete(apiRunInstance.id)
-      const currentCount = apisRunningByTraceId.get(apiRunInstance.traceId)
-      apisRunningByTraceId.set(apiRunInstance.traceId, currentCount - 1)
-
-      if (apisRunningById.size === 0) {
-        const { emitter } = require(`../redux`)
-        emitter.emit(`API_RUNNING_QUEUE_EMPTY`)
-      }
-
-      // Filter empty results
-      apiRunInstance.results = results.filter(result => !_.isEmpty(result))
-
-      // Filter out empty responses and return if the
-      // api caller isn't waiting for cascading actions to finish.
-      if (!args.waitForCascadingActions) {
-        apiSpan.finish()
-        resolve(apiRunInstance.results)
-      }
+  const { emitter } = require(`../redux`)
+  emitter.emit(`API_RUNNING_QUEUE_EMPTY`)
```

Comment:
What about scenario like this:

```js
exports.sourceNodes = async ({ actions }) => {

  await createNode(someNode)

  await new Promise(resolve {
    // simulate long running async operation
    setTimeout(() => {
      createNode(someOtherNode)
    }, 5000)
  })
}
```

The `API_RUNNING_QUEUE_EMPTY` would be emitted after creating `someNode` and running transformers for it - but `sourceNodes` is not completed yet and we shouldn't emit yet. I think we still need `apiRunningById` (or similar) tracker to make sure we emit only if there are no APIs running